File: v1/src/main/java/com/api/v1/services/customer/CustomerRegistrationServiceImpl.java
package com.api.v1.services.customer;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.api.v1.builders.customer.CustomerBuilder;
import com.api.v1.domain.entities.Customer;
import com.api.v1.domain.repositories.CustomerRepository;
import com.api.v1.dtos.requests.CustomerRegistrationRequestDto;
import com.api.v1.dtos.responses.CustomerResponseDto;
import com.api.v1.exceptions.customer.DuplicatedSsnException;
import com.api.v1.mappers.customer.CustomerResponseMapper;

import jakarta.validation.Valid;
import reactor.core.publisher.Mono;

@Service
class CustomerRegistrationServiceImpl implements CustomerRegistrationService {

    @Autowired
    private CustomerRepository repository;

    @Override
    public Mono<CustomerResponseDto> register(@Valid CustomerRegistrationRequestDto request) {
        return repository
            .findAll()
            .filter(customer -> customer.getSsn() != null
                    && customer.getSsn().equals(request.ssn())
                    && customer.getArchivedAt() == null
            )
            .hasElements()
            .flatMap(exists -> {
                if (exists) return Mono.error(new DuplicatedSsnException(request.ssn()));
                return Mono.defer(() -> {
                    Customer newCustomer = CustomerBuilder.create().fromDto(request).build();
                    Mono<Customer> savedCustomer = repository.save(newCustomer);
                    return savedCustomer.flatMap(customer -> Mono.just(CustomerResponseMapper.map(customer)));
                });
            });
    }
    
}

File: kotlin/v2/src/main/kotlin/com/api/v2/CustomerRegistrationServiceImpl.kt
package com.api.v2

import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class CustomerRegistrationServiceImpl: CustomerRegistrationService {

    @Autowired
    lateinit var customerRepository: CustomerRepository

    override suspend fun register(customer: @Valid Customer): CustomerResponseDto {

        suspend fun isSsnDuplicated() {
            if (customerRepository
                    .findAll()
                    .filter { e -> e.ssn == customer.ssn }
                    .count() != 0) {
                throw DuplicatedSsnException(customer.ssn)
            }
        }

        return withContext(Dispatchers.IO) {
            isSsnDuplicated()
            val savedCustomer = customerRepository.save(customer)
            CustomerResponseMapper.map(savedCustomer)
        }

    }


}